### PR TITLE
[docs] Add chat templates page to web docs

### PR DESCRIPTION
**Pull request by sergiopaniego**
# What does this PR do?

Follow-up to #5545, adding the content to the web docs

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@albertvillanova @qgallouedec 

---

> [!NOTE]
> **Low Risk**
> Documentation-only changes (new page, navigation entry, and links) with no runtime or API impact.
> 
> **Overview**
> Adds a new `Chat Templates` conceptual guide documenting TRL’s bundled/patched training chat templates (prefix-preserving + `generation` markers) and which model families are supported.
> 
> Wires the page into the docs nav (`_toctree.yml`) and adds cross-references from `chat_template_utils.md`, plus contextual notes in `grpo_trainer.md` (tools require prefix-preserving templates) and `sft_trainer.md` (assistant-only loss requires `generation` markers and points to the bundled template list).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6099d3c48e69316263239a28cca70263fc11224f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/source/chat_templates.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# Chat Templates
+
+TRL ships a small collection of Jinja2 chat templates under [`trl/chat_templates/`](https://github.com/huggingface/trl/tree/main/trl/chat_templates). They serve two purposes:
+
+1. **Identity comparison**: detecting which model is being used (by comparing `tokenizer.chat_template` against known templates) to add the appropriate response schema ([`add_response_schema`]) or swap in a training template ([`get_training_chat_template`]).
```
Comment (albertvillanova) [inline docs/source/chat_templates.md]:
```suggestion
1. **Identity comparison**: detecting which model is being used (by comparing `processing_class.chat_template` against known templates) to add the appropriate response schema ([`add_response_schema`]) or swap in a training template ([`get_training_chat_template`]).
```

We are indeed comparing `processing_class.chat_template` because although currently both processor and processor.tokenizer have equal `chat_template`, this is not enforced: either one could be modified independently, while the other could not.

Personally, this implementation in transformers does not make sense to me, but this is how it is. See my comment: https://github.com/huggingface/trl/pull/5489#discussion_r3098101928
> - Shouldn't they enforce that both (processor and tokenizer) chat_template's are the same?
> - Does it make sense that a tokenizer.response_schema is aligned with its processor.chat_template, but not with its own tokenizer.chat_template?

See:
- #5520

**Review by albertvillanova — APPROVED** [review]
Thanks! Just a small suggestion due to recent code changes...

File: docs/source/chat_templates.md (hunk 1 above)
Comment (sergiopaniego) [reply docs/source/chat_templates.md]: thanks!!

File: docs/source/chat_template_utils.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Chat template utilities
 
+For an overview of the chat templates bundled with TRL and the rationale behind the training patches, see [Chat Templates](chat_templates).
+
+## add_response_schema
```
Comment (qgallouedec) [inline docs/source/chat_template_utils.md]: I am not sure about adding this function in the doc. If we do this, it means that this function is meant to be used directly by the users. However, this function is more an internal utility in my view

Comment (sergiopaniego) [reply docs/source/chat_template_utils.md]: you're right, removed

File: docs/source/grpo_trainer.md [hunk 3]
```diff
@@ -632,6 +632,9 @@ trainer = GRPOTrainer(
 Each tool must be a standard Python function with **type-hinted arguments and return types**, along with a **Google-style docstring** describing its purpose, arguments, and return value.
 For more details, see the [Passing tools guide](https://huggingface.co/docs/transformers/en/chat_extras#passing-tools).
 
+> [!TIP]
+> The GRPO tool call loop requires the chat template to be *prefix-preserving* (appending messages must not change how earlier messages are rendered). For known model families (e.g. Qwen3, DeepSeek-V3), TRL automatically swaps in a patched training template when tools are enabled. See [Chat Templates](chat_templates#training-templates) for the full list.
```
Comment (qgallouedec) [inline docs/source/grpo_trainer.md]:
```suggestion
> The GRPO tool call loop requires the chat template to be *prefix-preserving* (appending a tool message must not change how earlier messages are rendered). For known model families (e.g. Qwen3, DeepSeek-V3), TRL automatically swaps in a patched training template when tools are enabled. See [Chat Templates](chat_templates#training-templates) for the full list.
```

a bit more precise

File: docs/source/chat_templates.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
+# Chat Templates
+
+TRL ships a small collection of Jinja2 chat templates under [`trl/chat_templates/`](https://github.com/huggingface/trl/tree/main/trl/chat_templates). They serve two purposes:
+
+1. **Identity comparison**: detecting which model is being used (by comparing `processing_class.chat_template` against known templates) to add the appropriate response schema ([`add_response_schema`]) or swap in a training template ([`get_training_chat_template`]).
```
Comment (qgallouedec) [inline docs/source/chat_templates.md:5]:
```diff
- ([`add_response_schema`])
+ (`add_response_schema`)
```

now that it's not in the doc anymore

File: docs/source/chat_templates.md
Comment (qgallouedec) [inline docs/source/chat_templates.md:1]: Nice!

I'd suggest a few modifications, flagging the rationale:

Reading the original intro cold, I felt it opened mid-thought: it jumped straight into "they serve two purposes: identity comparison and training patches", which is framed from the implementer's side (ie, why these files exist internally) rather than the reader's (=user) side: "why do I, a TRL user, care?". I reworked it along this arc:

- Define what a chat template is (+ link to transformers docs, tiny `apply_chat_template` example).
- Reassure: most users never touch them; TRL handles it transparently.
- Motivate the page with the two user-facing scenarios (SFT -assistant_only_loss=True`, GRPO tool calls).
- Say what to do: TRL auto-patches supported families; for others, you patch yourself.

Collapsed "Original templates" into "Supported model families". The per-file stubs ("Original Qwen3 chat template.", ...) weren't user-facing: those originals exist for TRL's internal identity-comparison; a user would never use them directly -> collapsed to a one-line family list

File: docs/source/chat_templates.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+Original Qwen3 chat template.
+
+### `qwen3_vl.jinja`
+
+Original Qwen3-VL chat template. Unlike text-only Qwen3, this template is already prefix-preserving (no conditional thinking blocks), so no training patch is needed.
+
+### `qwen3_5_2b_and_below.jinja` / `qwen3_5_4b_and_above.jinja`
+
+Original Qwen3.5 chat templates.
```
Comment (qgallouedec) [inline docs/source/chat_templates.md:50]:
````suggestion
A [chat template](https://huggingface.co/docs/transformers/en/chat_templating) is a Jinja2 snippet that formats messages into the string a model was trained on. For example:

```python
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
>>> tokenizer.chat_template
"{% for message in messages %}{% if loop.first and messages[0]['role'] != 'system' %}{{ '<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n' }}{% endif %}{{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}{% endfor %}{% if add_generation_prompt %}{{ '<|im_start|>assistant\n' }}{% endif %}"
>>> tokenizer.apply_chat_template([{"role": "user", "content": "Hi!"}], tokenize=False)
'<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nHi!<|im_end|>\n'
```

In most cases you don't need to worry about chat templates: models ship their template along with the tokenizer, and TRL applies it for you. The whole thing is transparent. But some TRL recipes rely on features that most shipped templates don't include:

- **SFT with `assistant_only_loss=True`** needs `&#123;% generation %&#125;` / `&#123;% endgeneration %&#125;` markers around assistant output, so the loss mask can target only assistant tokens.
- **GRPO with tool calls** needs the template to be *prefix-preserving*: appending a tool message must not change how earlier messages are rendered.

TRL ships patched templates under [`trl/chat_templates/`](https://github.com/huggingface/trl/tree/main/trl/chat_templates) for common families (Qwen, Llama, DeepSeek-V3, GPT-OSS, ...) and swaps them in automatically for supported models. For any other model, you'll need to patch its template yourself. The rest of this page catalogs what's bundled.

## Supported model families

TRL stores reference copies of the original templates so it can identify supported models at init and swap in a training template when needed. The following families are recognized: DeepSeek-V3, GLM-4-MoE, GPT-OSS, Llama 3 / 3.1 / 3.2, Qwen2.5, Qwen3, Qwen3-VL, Qwen3.5.
````

File: docs/source/chat_templates.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+### `qwen2_5_training.jinja`
+
+Patched Qwen2.5 template. Diff vs `qwen2_5.jinja`:
+
+Wrap assistant message output with `&#123;% generation %&#125;` / `&#123;% endgeneration %&#125;` so that `return_assistant_tokens_mask=True` produces correct masks for SFT assistant-only loss.
+
+## Related utilities
+
+See [Chat Template Utilities](chat_template_utils) for the helper functions (`clone_chat_template`, `is_chat_template_prefix_preserving`, `get_training_chat_template`) that operate on these templates.
```
Comment (qgallouedec) [inline docs/source/chat_templates.md:111]:
```suggestion
See [Chat Template Utilities](chat_template_utils) for the helper functions ([`clone_chat_template`], [`is_chat_template_prefix_preserving`], [`get_training_chat_template`]) that operate on these templates.
```

**Review by qgallouedec — APPROVED** [review]
